Fix Plone 4.3 with p.a.widgets 1.x edit modals

petschki opened the pull request:
see issue #719

petschki commented:
bah ... sorry for the mess in here ... tried to squash some commits, but gave up ... nevermind.

hvelarde commented on src/collective/cover/static/js/compose.js:
```diff
@@ -96,7 +96,9 @@ $(document).ready(function() {
 
   TitleMarkupSetup();
 
-  if ($.fn.prepOverlay !== undefined) {
```
changes on JS usually require an upgrade step to cook the resources; do you mind to check if we already have one in place?

petschki replied:
done that

hvelarde commented on src/collective/cover/widgets/richtext.py:
```diff
@@ -0,0 +1,50 @@
+import pkg_resources
+
+try:
```
I think this is not necessary as this code will only be execute if plone.app.widgets is installed and you already are checking for that on `configure.zcml`.

also, for me is pretty important to document what you're doing here, so we can remove this in the future when we drop compatibility with Plone 4. that sais, please add docstrings at the module and class levels explaining why this is needed..

petschki replied:
sure .. you're absolutely right ... i'll remove that

hvelarde replied:
we're still missing the docstrings.

petschki commented:
I'll try to fix the tests today ...

petschki commented:
@hvelarde everything is green except one test in DEXTERITY_ONLY setup ... and I really do not know whats going wrong there. Can you give me some pointers? Could I add this test to the noncritical list (like `issue_637` already is)?

hvelarde commented:
@petschki those are mandelbugs; I restarted the builds for you; what happened with the idea of working on this directly in plone.app.widgets?

petschki commented:
@hvelarde  ah ... thanks for restarting. Now we're green. But you're right about RichTextWidget... let me check p.a.widget and the pull request https://github.com/plone/plone.app.widgets/pull/160 before merging this...

petschki commented:
WIP: https://github.com/plone/plone.app.widgets/pull/163

petschki commented:
https://github.com/plone/plone.app.widgets/pull/163 was merged right now ... I'll request a release of p.a.widgets and cleanup the unused code here.

hvelarde commented:
do we still to add something in collective.cover, or just using the new widgets solve your problem?

petschki commented:
@hvelarde unfortunately we still have to map a custom FieldWidget adapter for the RichTextField (like its done in plone.app.widgets.dx_bbb) ... see https://github.com/plone/plone.app.widgets/issues/164 ... my latest pull request would fix this, so everything would work supermagically https://github.com/plone/plone.app.widgets/pull/165

hvelarde commented:
ok, just keep me updated :-)

hvelarde commented:
@petschki do you mind to make a rebase? also, do you mind to update me the status on this after the changes you made on plone.app.widgets?

petschki commented:
@hvelarde sure ... I'll rebase later today. the `plone.app.widgets` progress made the integration a bit easier, so we do not need to wire any widgets in collective.cover anymore. right now, it depends on the latest `1.x` branch but I think we can have a p.a.widgets release 1.10 this week. I've also setup a testing environment for the widgets implementation but got stuck in robots test, not finding elements within the plone-modal ... need to do more investigations on that this week. I'll commit the `4.3.x-widgets` travis-environment later today and put it to allow_failures ...

hvelarde commented:
we're having some mandelbugs with RF tests right now and I need to restart failing builds from time to time.

I think the problem on the new build you just created seems to be related with a feature removed from zc.buildout, but I don't remember very well and I'm not sure.

FWIW, buildout previously was able to downgrade itself and I think that's not the case anymore; that's why you're having:

```python
VersionConflict: (zc.buildout 2.9.4 (/home/travis/build/collective/collective.cover/eggs/zc.buildout-2.9.4-py2.7.egg), Requirement.parse('zc.buildout==1.7.1'))
```

probably @idgserpro knows what is happening here.

idgserpro commented:
We have good news and bad news. Let's start with the good ones, the reason (and a possible workaround) of the error.

When you use bootstrap.py, it downloads the latest zc.buildout (that now is 2.9.4). In `bin/buildout` it will point to `/home/travis/build/collective/collective.cover/eggs/zc.buildout-2.9.4-py2.7.egg`.

Here is the insanity: after bootstraping, when running `bin/buildout`, when **downloading** and extracting zc.buildout 1.7.1 for the *first* time when it's a direct dependency (`install_requires`) in an extensions section, it will give you the error: but if you already have a `/home/travis/build/collective/collective.cover/eggs/zc.buildout-1.7.1-py2.7.egg` in place, the buildout continues. That's why the error was happening only in the new job with the mr.developer in extensions, the old ones don't have mr.developer and had a cache that the new job didn't.

You can simulate it in your machine instead of travis, in a clean virtualenv environment (remember to comment a `eggs-directory` in `~/.buildout/default.cfg` if you have it in place). Funny thing is: the first time you run it will break (giving the error), but the second time won't, because you now have a zc.buildout-1.7.1-py2.7.egg in your environment.

The bad news: It seems that it started in zc.buildout 2.9.x branch (https://travis-ci.org/collective/collective.cover/builds/247259422) but we don't have time to check exactly what happened to start having this issue. As miohtama used to say, *life is hard and Plone is harder™*.

`setuptools` have the same issue as zc.buildout above: travis decides to [automatically install some packages](https://docs.travis-ci.com/user/languages/python/#Pre-installed-packages) and `setuptools` is a direct dependency from some of them (the specific versions can be [checked here](https://travis-ci.org/collective/collective.cover/builds/247295792#L150)).

There are two immediate workarounds:

- Pin zc.buildout and setuptools to the last version in versions-4.3.x-widgets.cfg (https://travis-ci.org/collective/collective.cover/jobs/247501030);
```
# VersionConflict: (zc.buildout 2.9.4 (/home/travis/build/collective/collective.cover/eggs/zc.buildout-2.9.4-py2.7.egg), Requirement.parse('zc.buildout==1.7.1'))
# We still don't know why this happens. It seems that when using extensions with a package that
# requires zc.buildout (like mr.developer).
zc.buildout = 
# XXX: Same as above.
# VersionConflict: (setuptools 12.0.5 (/home/travis/virtualenv/python2.7.9/lib/python2.7/site-packages), Requirement.parse('setuptools==20.1.1'))
# travis has some packages installed by default in it's virtualenv and setuptools is installed.
# https://docs.travis-ci.com/user/languages/python/#Pre-installed-packages
setuptools =
```
- Run bootstrap with Plone 4.3 pins of zc.buildout and setuptools, `bootstrap.py --buildout-version=1.7.1 --setuptools-version=20.0.1` in `.travis.yml` (https://travis-ci.org/collective/collective.cover/jobs/247520681)
```
# XXX:
# Trying to pin in versions-4.3.x-widgets to the latest version it works but then all tests break.
# So we're going to use the same versions that Plone uses in it's versions.
- PLONE_VERSION=4.3 WIDGETS=true BOOTSTRAP_PARAMS="--buildout-version=1.7.1 --setuptools-version=20.1.1"
(...)
python bootstrap.py $BOOTSTRAP_PARAMS
```

Bad news part 2: now you have **20** failed tests after solving the `VersionConflict`. @petschki, can you try to run locally and see if this happens only in Travis?

Another question: why are you pinning `plone.app.widgets = ` since you're already checking it's branch in `sources`?

petschki commented:
@idgserpro thanks for the details ... pinned everything down and removed unneeded version pin of plone.app.widgets. not the buildout runs smooth ... I'm aware of those 20 errors. thats why I moved the environment to allow_failures ..

idgserpro commented:
You don't need the two pins, only one of them. It's a good practice as well to document the pins in the cfg itself, so in the future if you try to remove them its less work to know why it was pinned. Our suggestion to doc above the pins:

```
# XXX: https://github.com/collective/collective.cover/pull/720#issuecomment-311383938
```

hvelarde commented:
I not sure which one would be best practice here; I guess we should ask.

idgserpro commented:
I think we should open an issue at zc.buildout exposing the problem and, from there, we can know which one is the best approach.

From a Plone perspective, since they are planning to get rid of bootstrap.py, they will problably be in favour of changing cfgs.

rodfersou commented on src/collective/cover/static/js/compose.js:
```diff
@@ -96,8 +96,18 @@ $(document).ready(function() {
+    if($("body").hasClass("pat-plone-widgets")) {
+      // XXX: reload tile content asynchronously here when destroying the modal
+      //      need to find the right event where to hook in
+      return  // exit here
```
miss semicolon here

hvelarde reviewed (changes requested):
the branch is outdated; a review is not possible in this state.
